vertical-pod-autoscaler/pkg/recommender/routines/replica_restrictions_post_processor.go

```diff
@@ -0,0 +1,208 @@
+	replicas int,
+	lowWatermark, highWatermark int) (*vpa_types.RecommendedContainerResources, error) {
+
+	amendedRecommendations := containerRecommendation.DeepCopy()
+
+	isRestricted := isRestrictedScaling(containerRecommendation, container, lowWatermark, highWatermark, replicas)
+	process := func(recommendation apiv1.ResourceList) {
+		applyRestrictedScalingVPAPolicy(recommendation, isRestricted, container.Resources.Requests)
+	}
+
+	process(amendedRecommendations.Target)
+	// Do we want to change LowerBound and UpperBound?
+
+	return amendedRecommendations, nil
+}
+
+// isRestrictedScaling: detects if the recommendation is a restricted scaling
+func isRestrictedScaling(containerRecommendation *vpa_types.RecommendedContainerResources, container apiv1.Container, lowWatermark, highWatermark int, replicas int) bool {
+	isAnUpscale := isUpscale(containerRecommendation.Target, container.Resources.Requests)
+	isADownscale := isDownscale(containerRecommendation.Target, container.Resources.Requests)
+	isRestricted := false
+	if isAnUpscale && replicas <= highWatermark {
+		isRestricted = true
+	}
+	if isADownscale && replicas >= lowWatermark {
+		isRestricted = true
+	}
+	if isRestricted {
+		klog.V(2).Infof("[%s] Scaling is restricted %d - replica restrictions [%d, %d] - upscale: %t - downscale: %t", container.Name, replicas, lowWatermark, highWatermark, isAnUpscale, isADownscale)
+	} else {
+		klog.V(4).Infof("[%s] Scaling is not restricted %d - replica restrictions [%d, %d] - upscale: %t - downscale: %t", container.Name, replicas, lowWatermark, highWatermark, isAnUpscale, isADownscale)
+	}
+	return isRestricted
+}
+
+// isUpscale: detects if the recommendation is an upscale. Any increase of resources is considered as upscale.
+func isUpscale(recommendation apiv1.ResourceList, containerOriginalResources apiv1.ResourceList) bool {
+	for recommendationType, _ := range recommendation {
+		q := containerOriginalResources.Name(recommendationType, resource.DecimalSI).MilliValue()
+		r := recommendation.Name(recommendationType, resource.DecimalSI).MilliValue()
+		if r > q {
+			return true
+		}
+	}
+	return false
+}
+
+// isDownscale: detects if the recommendation is a downscale. Any decrease of resources is considered as downscale.
+func isDownscale(recommendation apiv1.ResourceList, containerOriginalResources apiv1.ResourceList) bool {
+	for recommendationType := range recommendation {
+		q := containerOriginalResources.Name(recommendationType, resource.DecimalSI).MilliValue()
+		r := recommendation.Name(recommendationType, resource.DecimalSI).MilliValue()
+		if r < q {
+			return true
+		}
+	}
+	return false
+}
+
+// applyRestrictedScalingVPAPolicy makes sure we do not make a recommendation if we are restricted.
+func applyRestrictedScalingVPAPolicy(recommendation apiv1.ResourceList, isRestricted bool, containerOriginalResources apiv1.ResourceList) {
+
+	for recommendationType := range recommendation {
+		if isRestricted {
```

Comment:
While I think code is correct, I cannot understand how this solution is going to be stable.
Based on an instant value that is the replica count (that is influenced by horizontal scaling) you can suddenly go back to chart initial value! While the jump might me contained with the `triggered` mode due to the chart mutation, this can work well with classic mode `auto` and `recreate`.
I can see the recommendation slowly moving from an initial value toward a target, potentially with big difference, and suddenly going back to what the chart was defining.

Reply:
https://datadoghq.atlassian.net/wiki/spaces/AG/pages/2971205765/WPA+VPA+Cooperation#Iteration-0

Yes we should not look at instant metrics in the final solution, but it's probably still good enough to start since services usually don't jump from 0 to 100 pods and back to 0 every 10 seconds

Reply:
fixed/mitigated thanks to check in the process function `*vpa.UpdateMode != vpa_types.UpdateModeTrigger`